### ISSUE TITLE
DL4J: Improve performance of DL4J with MKL-DNN for ResNet50

### ISSUE THREAD

**Issue by dollarHome** [deeplearning4j/deeplearning4j#7521]
#### Issue Description
Attention: @AlexDBlack 

**This issue would be applicable to any NN model (which uses Pooling/BN) for Inference**
ResNet50 with DL4J on 28c Xeon system is not scaling to the cores. I debugged this issue further and with MKLDNN_VERBOSE=1, and you can see the reason why (see mkldnn log). As per MKLDNN team, as discussed here: https://github.com/intel/mkl-dnn/issues/406 reorders around pooling and batchnorm are inducing additional overheads in the computation.  Quoting MKLDNN team: "For inference there's no need to reorder weights all the time as these do not change." "These should be reordered once when model is loaded for instance and kept in the optimized layout."

This is after all the platform tuning flags set: 
* SKX 28 Cores (JVM pinned to 1 socket)
* Heap size: 29GB, System set to Performance mode, Hyperthreading ON, Total imagenet images used in inference: 2476
**Platform tuning flags used:**
KMP_BLOCK_TIME=0, OMP_WAIT_POLICY=PASSIVE, MKL_THREADING_LAYER=GNU, MKL_NUM_THREADS=28, OMP_NUM_THREADS=28

**Please describe our issue, along with:**
[SimpleWorkload_MKLDNN_VERBOSE_snipped.log](https://github.com/deeplearning4j/deeplearning4j/files/3066024/SimpleWorkload_MKLDNN_VERBOSE_snipped.log)

- Exepected Behavior: Performance would scale to the cores
- Actual Behavior: Unnecessary overheads introduced due to MKLDNN reordering (which is not necessary for Inference). Performance is not scaling with increased batchsize and cores.

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version - snapshot
* platform information (OS, etc) - Linux, Ubuntu 18.04

#### Contributing
Yes, I would like to help fix this issue.

**Comment by AlexDBlack** [comment]
This will be non-trivial to implement; eventually however we should look at it (and look at it for CuDNN also).

We would need 2 things here:
1. Support for different weight forwards for conv etc layers.
2. Automatic rearrangement if required

On the second point, I don't think it'll be feasible to leave the weight layout decision up to users. Not sure if optimal weight layout will be different for cuDNN also.

Because the DL4J conv layer is designed and optimized around the existing input/weight layouts, I would suggest we look at this once the CUDA branch is merged, so we can switch the underlying implementation to conv2d op (that is more flexible for input/weight layouts).
https://github.com/deeplearning4j/deeplearning4j/pull/7095

**Comment by saudet** [comment]
What MKL-DNN uses exactly is pretty opaque, it's not something we should be using for storage and is almost certainly different than what cuDNN uses internally. What we could do about it though is have a cache with the weights in the correct order when we're in inference mode. And to generalize for training, we could have a way to sync it only when we need the weights outside MKL-DNN.

**Comment by AlexDBlack** [comment]
External cache for inference makes sense, though we'll need to be careful with the extra memory use for that.
As for training: we always need to sync it back to the full parameter vector on each iteration (for updating, and other parts of SGD) so that's unfortunately not an option.

**Comment by dollarHome** [comment]
@AlexDBlack  I understand this can be a complex task. Please let me know how I can help here. I am looking at the code too. A little more detailed explanation with code pointers can put me in the right direction.

**Comment by saudet** [comment]
@dollarHome I think we'll need a cache that's going to happen mostly in C++. For example, the code for 2d convolution happens here: https://github.com/deeplearning4j/deeplearning4j/blob/master/libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp#L602-L678
You'll see that on init (everything after `checkAndReset()` but before `submitAndWait()`) it creates the stream and all. But as part of that init, we could also build a second stream just for the reorder, and run that one only once. This is something we could also enhance later by deallocating the original buffer when it's not needed anymore.
